README.md

```diff
@@ -1,73 +1,124 @@
-  "createdAt": ""               // user created date
-}
+
+To test with production build, use `docker-compose` then it will launch necessary service and config for production:
+
+```sh
+docker-compose build
+docker-compose up
 ```
 
-## Requirements
+## Guideline for feature development
 
-### Functionality
+As a backend engineer, consider three kinds of characteristic:
 
-- The API should follow typical RESTful API design pattern.
-- The data should be saved in the DB.
-- Provide proper unit test.
-- Provide proper API document.
+1. How we design a RESTful interface?
+   - [MicrosoftREST API Guideline](https://github.com/microsoft/api-guidelines/blob/vNext/Guidelines.md) is a good reference.
+   - Idempotency is the key to make RESTful API useful and reliable.
+2. How we design domain model?
+   - In domain tier: Represent domain's properties and behaviours.
+   - In application tier: Represent use cases and exceptions.
+3. How we persist data in database?
+   - Estimate size of data, cost to backup, time to restore as much as possible.
+   - Ensure [each online query uses index](https://docs.mongodb.com/manual/tutorial/measure-index-use/).
+   - If we assume high-load access on API, consider to apply [CQRS](https://docs.microsoft.com/en-us/azure/architecture/patterns/cqrs), [Event Sourcing](https://docs.microsoft.com/en-us/azure/architecture/patterns/event-sourcing) or [other data management pattern](https://docs.microsoft.com/en-us/azure/architecture/patterns/category/data-management) for better performance.
 
-### Tech stack
+### Case 1: Friends feature
 
-- Use Node.js and any framework.
-- Use any DB. NoSQL DB is preferred.
+When we have a new requirement like "followers/following" or "friends" feature, consider following requirements:
 
-### Bonus
+- How many friends a user can have at most.
+  - If it's many, we cannot use [Embedded Data Models](https://docs.mongodb.com/manual/core/data-model-design/#embedded-data-models) so we need to consider deeper about consistency.
+- When a user adds a friend, how long time we can wait until the target's followers have updated.
+  - If we can wait seconds, [event-driven architecture](https://docs.microsoft.com/en-us/azure/architecture/guide/architecture-styles/event-driven) could be better for resiliency.
+  - [Transaction](https://docs.mongodb.com/manual/core/transactions/) is also an option but [it probably reduce our options in operation phase like shading](https://www.mongodb.com/blog/post/performance-best-practices-transactions-and-read--write-concerns).
 
-- Write clear documentation on how it's designed and how to run the code.
-- Write good in-code comments.
-- Write good commit messages.
-- An online demo is always welcome.
+When code, The 'follow' is `User`'s behaviour, so implement a `follow(User)` method in `User` class like:
 
-### Advanced requirements
+```typescript
+private readonly friends: Set<UserId>;
+private readonly followers: Set<UserId>;
```

Comment:
If we have a need to get followers of a specific user with the order by `followed time`, how do you manage to design the data structure for `followers`?

Reply:
The easiest idea is to replace this `Set` with `List`, and add new friends/followers to the end of these lists.
It would be enough if we just care about its order. We need less change even on the datastore side.

If we need to persist the `followed time` record, we can use `List<{userId: UserId, date: Date}>` in the domain tier.  
To design its data model for mongo, we have several ideas:

1. [Embedded data model](https://docs.mongodb.com/manual/core/data-model-design/#embedded-data-models), if and only if followers are enough little and we're sure that document size is smaller than [the maximum BSON document size](https://docs.mongodb.com/manual/reference/limits/#mongodb-limit-BSON-Document-Size).
2. [Normalized data model](https://docs.mongodb.com/manual/core/data-model-design/#normalized-data-models) with an isolated document that persists necessary props such as `user`, `follower`, and `created time`. The [compound index](https://docs.mongodb.com/manual/core/index-compound/) helps us to find documents by `user` and sort by `created time`.